src/LagrangianParticleTracking/LagrangianParticleTracking.jl

```diff
@@ -0,0 +1,66 @@
+module LagrangianParticleTracking
+
+export LagrangianParticles, advect_particles!
+
+using Adapt
+using KernelAbstractions
+using StructArrays
+
+using Oceananigans.Grids
+using Oceananigans.Architectures: device
+using Oceananigans.Fields: interpolate, datatuple, compute!, location
+using Oceananigans.Utils: MAX_THREADS_PER_BLOCK
+
+import Base: size, length, show
+
+abstract type AbstractParticle end
+
+struct Particle{T} <: AbstractParticle
+    x :: T
+    y :: T
+    z :: T
+end
+
+struct LagrangianParticles{P, R, T}
+         particles :: P
+       restitution :: R
+    tracked_fields :: T
+end
```

Comment:
For the sake of future dev's and myself and also the docs, do you mind providing some comments / docs on the purpose of the `LagrangianParticles` struct? Naively I would think it would make sense to store everything we need particle-wise so that `model.particles` would just be a simple `Array` / `CuArray`. I'm sure there's a good reason for this decision, just can't detect it at a glance.

Reply:
I added some model setup docs for Lagrangian particle tracking showing how to track simple particles and particles with custom properties (+ writing particle properties to JLD2/NetCDF).

As for the purpose of the `LagrangianParticles` struct: Initially I just kept track of the particle's locations/properties so a single `StructArray` was enough. But then it became obvious that there would be some properties shared by all particles, e.g. a coefficient of restitution and which fields are being tracked. There may be more in the future if we add more features.

The coefficient of restitution could be stored by each particle but that would increase memory requirements for the narrow case of wanting a different coefficient for each particle.

Reply:
> But then it became obvious that there would be some properties shared by all particles, e.g. a coefficient of restitution and which fields are being tracked.

Why do we want to enforce that all particles have the same coefficient of restitution? Isn't this a limitation?

Reply:
I would advocate for an implementation that's as flexible as possible, even if it does impose some extra memory or computational requirements. I think the new science afforded by a flexible abstraction might be more important than faster simulations or the ability to run with huge amounts of particles, right? We can always provide "optimized" implementations if that becomes important in the future; but we might not want to limit what people can do for performance whose necessity is unclear?

Reply:
The current implementation is flexible enough to do almost anything.

You can assign different coefficients of restitution to each particle and implement any kind of particle-wall collision you want.

See #1244 for an example of using custom particle types to implement particle-particle interactions.

But yes in this PR by default there is one coefficient of restitution for all particles. If you think it's important that each particle has it's own coefficient I'm happy to make the change if it means the PR can be merged.